sendrecv/gst-rust/src/main.rs

```diff
@@ -0,0 +1,768 @@
+                    let mbuilder = gst::Message::new_application(gst::Structure::new(
+                        "ws-message",
+                        &[("body", &msg)],
+                    ));
+                    let _ = bus.post(&mbuilder.build());
+                }
+
+                _ => {
+                    println!("Unmatched message type: {:?}", message);
+                }
+            }
+        }
+    })
+}
+
+fn handle_application_msg(
+    app_control: &AppControl,
+    struc: &gst::StructureRef,
+) -> Result<(), Error> {
+    match struc.get_name() {
+        "ws-message" => {
+            let msg = struc.get_value("body").unwrap();
+            app_control.clone().on_message(msg.get().unwrap())
+        }
+        "ws-error" => Err(WsError(app_control.0.lock().unwrap().app_state.clone()))?,
+        "error" => {
+            let msg: String = struc.get_value("body").unwrap().get().unwrap();
+            Err(BusError(msg))?
+        }
+        _u => {
+            println!("Got unknown application message {:?}", _u);
+            Ok(())
+        }
+    }
+}
+
+fn main() {
+    gst::init().unwrap();
+    match check_plugins() {
+        Err(err) => {
+            println!("{:?}", err);
+            return;
+        }
+        _ => {}
+    }
+
+    let (server, peer_id) = parse_args();
+
+    println!("Connecting to server {}", server);
+    let client = match websocket::client::ClientBuilder::new(&server)
+        .unwrap()
+        .connect_insecure()
+    {
+        Ok(client) => client,
+        Err(err) => {
+            println!("Failed to connect to {} with error: {:?}", server, err);
+            return;
+        }
+    };
+
+    let main_loop = glib::MainLoop::new(None, false);
+    let pipeline = gst::Pipeline::new("main");
+    let bus = pipeline.get_bus().unwrap();
+    let (receiver, sender) = client.split().unwrap();
+    let (send_msg_tx, send_msg_rx) = mpsc::channel::<OwnedMessage>();
+
+    let send_loop = send_loop(sender, send_msg_rx);
+
+    let bus_clone = bus.clone();
+    let send_msg_tx_clone = send_msg_tx.clone();
+    let receive_loop = receive_loop(receiver, send_msg_tx_clone, bus_clone);
+
+    let app_control = AppControl(Arc::new(Mutex::new(AppControlInner {
+        webrtc: None,
+        pipeline: pipeline,
+        send_msg_tx: send_msg_tx,
+        bus: bus.clone(),
+        main_loop: main_loop.clone(),
+        peer_id: peer_id.to_string(),
+        app_state: AppState::ServerConnected,
+    })));
+    app_control.register_with_server();
+
+    bus.add_watch(move |_, msg| {
+        use gst::message::MessageView;
+        match msg.view() {
+            MessageView::StateChanged(_) => {}
+            MessageView::StreamStatus(_) => {}
+            MessageView::Latency(_) => {}
+            MessageView::AsyncDone(_) => {}
+            MessageView::Element(_) => {}
+            MessageView::Tag(_) => {}
+            MessageView::NewClock(_) => {}
+            MessageView::Warning(warning) => {
```

Comment:
You also want to handle `Error` messages, but all the others above are not really needed :) An `Error` message is always fatal